include/range/v3/action/slice.hpp

```diff
@@ -36,47 +36,100 @@ namespace ranges
-                        typename D = range_difference_type_t<Rng>>
-                    auto requires_() -> decltype(
-                        concepts::valid_expr(
-                            concepts::model_of<concepts::ForwardRange, Rng>(),
-                            concepts::model_of<concepts::ErasableRange, Rng, I, I>(),
-                            concepts::model_of<concepts::ConvertibleTo, T, D>(),
-                            concepts::model_of<concepts::ConvertibleTo, U, D>()
-                        ));
-                };
 
-                template<typename Rng, typename T, typename U>
-                using Concept = concepts::models<ConceptImpl, Rng, T, U>;
+              // Overloads for the pipe syntax: rng | view::slice(from,to)
+              template<typename D, CONCEPT_REQUIRES_(Integral<D>())>
+              static auto bind(slice_fn slice, D from, D to)
+              RANGES_DECLTYPE_AUTO_RETURN
+              (
+                  std::bind(slice, std::placeholders::_1, from, to)
+              )
+              template<typename D, CONCEPT_REQUIRES_(Integral<D>())>
+              static auto bind(slice_fn slice, D from, detail::from_end_<D> to)
+              RANGES_DECLTYPE_AUTO_RETURN
+              (
+                  std::bind(slice, std::placeholders::_1, from, to)
+              )
+              template<typename D, CONCEPT_REQUIRES_(Integral<D>())>
+              static auto bind(slice_fn slice, detail::from_end_<D> from, detail::from_end_<D> to)
+              RANGES_DECLTYPE_AUTO_RETURN
+              (
+                  std::bind(slice, std::placeholders::_1, from, to)
+              )
+              template<typename D, CONCEPT_REQUIRES_(Integral<D>())>
+              static auto bind(slice_fn slice, D from, const end_fn &to)
+              RANGES_DECLTYPE_AUTO_RETURN
+              (
+                  std::bind(slice, std::placeholders::_1, from, to)
+              )
+              template<typename D, CONCEPT_REQUIRES_(Integral<D>())>
+              static auto bind(slice_fn slice, detail::from_end_<D> from, const end_fn &to)
+              RANGES_DECLTYPE_AUTO_RETURN
+              (
+                  std::bind(slice, std::placeholders::_1, from, to)
+              )
+            public:
 
-                // TODO support slice from end.
-                template<typename Rng,
-                    typename I = iterator_t<Rng>,
-                    typename D = range_difference_type_t<Rng>,
-                    CONCEPT_REQUIRES_(Concept<Rng, D, D>())>
-                Rng operator()(Rng && rng, range_difference_type_t<Rng> from,
-                    range_difference_type_t<Rng> to) const
+              template<typename Rng,
+                  typename I = iterator_t<Rng>,
+                  typename D = range_difference_type_t<Rng>,
+                  CONCEPT_REQUIRES_(ForwardRange<Rng>() && ErasableRange<Rng, I, I>())>
+              Rng operator()(Rng &&rng, D from, D to) const
```

Comment:
Im afraid I did not fully understood the last comment.
Could you please expand on what that wrapper would do

Reply:
Ideally, it would `RANGES_EXPECT(detail::can_represent<range_difference_type_t<Rng>>(x))` where `x` is `from` or `to` (unwrapped if they're detail::from_end_`) and `detail::can_represent` is a range-v3 port of `__can_represent` from cmcstl2 (https://github.com/CaseyCarter/cmcstl2/blob/master/include/stl2/detail/algorithm/is_permutation.hpp#L30), and then invoke the `slice_fn` call operator with the values converted to `range_difference_type_t<Rng>` (or `from_end<range_difference_type_t<Rng>>`).